coresimd/arm/mod.rs

```diff
@@ -7,14 +7,21 @@
 //! http://infocenter.arm.com/help/topic/com.arm.doc.
 //! ihi0073a/IHI0073A_arm_neon_intrinsics_ref.pdf
 //! [arm_dat]: https://developer.arm.com/technologies/neon/intrinsics
+#![allow(non_camel_case_types)]
 
 mod v6;
 pub use self::v6::*;
 
+#[cfg(any(target_arch = "aarch64", target_feature = "v7"))]
 mod v7;
+#[cfg(any(target_arch = "aarch64", target_feature = "v7"))]
 pub use self::v7::*;
 
-#[cfg(target_feature = "neon")]
+// NEON is supported on AArch64, and on ARM when built with the v7 and neon
+// features. Building ARM without neon produces incorrect codegen.
```

Comment:
I guess this is probably related to the above comment, but what's the incorrect codegen here? Is that an LLVM/rustc bug?

Reply:
This might be an LLVM/rustc bug AFAICT.

On AArch64 we can compile without enabling the `neon` feature globally, use `#[target_feature(enable="neon")]` to enable neon for some functions, and call these with run-time feature detection without issues. The code generated looks pretty good.

ARMv6 and older do not support neon at all, so for ARM neon is only allowed if the `v7` feature is enabled (probably should check `v8` here as well but that's not supported by rustc yet?). 

Using the armv7-... target (with the v7 feature enabled) and tagging functions with `#[target_feature(enable="v7,neon")]` and calling them using run-time feature detection did work, but it produced horrible code (as in, way too many instructions). Enabling `neon` at compile-time produced the code I expected. So... maybe there is an LLVM/rustc bug where using`#[target_feature(enable = "v7,neon")]` is not enough for armv7 targets to generate good code? Maybe we need to enable some more features for these, I don't know yet.

Note that before, these modules where guarded with `#[cfg(target_feature = "neon")]`, so if these weren't enabled at compile-time we were not including them at all. Now it got a bit better for ARM, but I'd like to know what's going on with `neon` here.

Reply:
Hm ok, so it sounds like v7 is very different from AVX in x86_64, but it's more like the 64-bit intrinsics on x86_64 wrt the i686 platform? (fundamentally not possible).

If that's the case, do you have some examples of the v7+neon code generation that are bad? I'd be curious to dig in there and see what's going on!

Reply:
> Hm ok, so it sounds like v7 is very different from AVX in x86_64, but it's more like the 64-bit intrinsics on x86_64 wrt the i686 platform? (fundamentally not possible).

I think `arm`, `armv7`, and `aarch64` are just different targets, as different at least as `x86` and `x86_64`.

> If that's the case, do you have some examples of the v7+neon code generation that are bad? I'd be curious to dig in there and see what's going on!

So  branched this, removed those checks, and disabled neon in ci/run.sh on this branch (https://github.com/gnzlbg/stdsimd/commit/9c9b2cf01c038e0f5c99e97e16fa3ebbee76fbab), which results in this travis failure(https://travis-ci.org/gnzlbg/stdsimd/jobs/355738848#L3159) .

Now I remember what was wrong: the run-time tests were not passing! The results generated were incorrect. This might well be an LLVM bug.

Reply:
cc @petrochenkov @parched @Amanieu 

---

This also might be related to #139 where it was discussed that we might need to treat arm, armv7, armv8, as different architectures, some of which have a 64-bit mode.